From 784762f7fc514cb35d511e0929d6faa7ffacb45e Mon Sep 17 00:00:00 2001
From: Raymond Wong <infiwang@pm.me>
Date: Thu, 16 Feb 2023 02:44:29 +0800
Subject: [PATCH 046/156] riscv(emit): fix delta

MCode pointer arithmetic is 4-byte aligned, cast to 1-byte aligned
type.

This is a WIP, archive only.
---
 src/lj_emit_riscv.h | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/lj_emit_riscv.h b/src/lj_emit_riscv.h
index 8306e200..d33bdd43 100644
--- a/src/lj_emit_riscv.h
+++ b/src/lj_emit_riscv.h
@@ -180,6 +180,7 @@ static void emit_ext(ASMState *as, RISCVIns riscvi, Reg rd, Reg rs1)
         break;
       default:
         lj_assertA(0, "invalid ext op");
+        return;
     }
     emit_dsshamt(as, sri, rd, rd, shamt);   
     emit_dsshamt(as, sli, rd, rs1, shamt);
@@ -277,7 +278,7 @@ typedef MCode *MCLabel;
 static void emit_branch(ASMState *as, RISCVIns riscvi, Reg rs1, Reg rs2, MCode *target)
 {
   MCode *p = as->mcp;
-  ptrdiff_t delta = target - (p - 2);
+  ptrdiff_t delta = (char *)target - (char *)(p - 2);
   // lj_assertA(((delta + 0x10000) >> 13) == 0, "branch target out of range"); /* B */
   lj_assertA(((delta + 0x100000) >> 21) == 0, "branch target out of range"); /* ^B+J */
   if (checki13(delta)) {
@@ -293,7 +294,7 @@ static void emit_branch(ASMState *as, RISCVIns riscvi, Reg rs1, Reg rs2, MCode *
 static void emit_jmp(ASMState *as, MCode *target)
 {
   MCode *p = as->mcp;
-  ptrdiff_t delta = target - (p - 1);
+  ptrdiff_t delta = (char *)target - (char *)(p - 1);
   // lj_assertA(((delta + 0x100000) >> 21) == 0, "jump target out of range"); /* J */
   lj_assertA(checki32(delta), "jump target out of range"); /* AUIPC+JALR */
   if (checki21(delta)) {
@@ -312,7 +313,7 @@ static void emit_jmp(ASMState *as, MCode *target)
 static void emit_call(ASMState *as, void *target, int needcfa)
 {
   MCode *p = as->mcp;
-  ptrdiff_t delta = (char *)target - ((char *)(p - 1));
+  ptrdiff_t delta = (char *)target - (char *)(p - 1);
   if (checki21(delta)) {
     *--p = RISCVI_JAL | RISCVF_D(RID_RA) | RISCVF_IMMJ(delta);
   } else if (checki32(delta)) {
-- 
2.42.0

